### PR TITLE
Rewrite the etcd doc to be about upgrading etcd.

### PR DESCRIPTION
**Pull request by mml**
First draft, nearly verbatim copy.  Please rip to shreds.

---
This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes.github.io/2767)

### CODE REVIEW

**Comment by wojtek-t** [comment]
Just some minor comments.

**Comment by wojtek-t** [comment]
@mml ping

File: docs/admin/etcd.md [hunk 1]
```diff
@@ -1,49 +1,163 @@
...
+v2 API.  Also, we have no pre-existing process for updating etcd, as we have
+never updated etcd by either minor or major version.
+
+Note that we need to migrate both: etcd versions that we are using (from 2.2.1
+to at least 3.0.x) as well as the version of API Kubernetes talks to etcd (3.0.x
+etcd binaries support both v2 and v3 API).
+
+This document describes how to do this migration.
+
+### Etcd upgrade limitations
```
Comment (xiang90) [inline docs/admin/etcd.md]: etcd upgrade limitations. (do not use Etcd :P)

File: docs/admin/etcd.md [hunk 2]
```diff
@@ -1,49 +1,163 @@
...
+
+The way etcd was designed introduces some significant limitations on how the
```
Comment (xiang90) [inline docs/admin/etcd.md]: i do not think this is a design issue to be honest. we have done quite some internal work to allow removing these limitations from very beginning. we just do not know the scope of the use case and the effort. We should resolve it here:

https://github.com/coreos/etcd/issues/7308

File: docs/admin/etcd.md [hunk 3]
```diff
@@ -1,49 +1,163 @@
...
+cluster for any intermediate minor release, wait until it is healthy and
+functional and stop it then. This will do the migration itself underneath. 
+
+As an example: to upgrade from 2.1.x to 2.3.y version, it is enough to start
+etcd in 2.2.z version, wait until it is healthy, stop it and then start then
+2.3.y version, wait until it is up and we are done.
+
+#### No rollback
+
+Etcd doesn’t support rollback procedure at all (by design). That means, that if
```
Comment (xiang90) [inline docs/admin/etcd.md]: etcd

**Comment by xiang90** [comment]
This doc looks good to me in general.

etcd team is OK to support rollback once https://github.com/coreos/etcd/issues/7308 is resolved.

File: docs/admin/etcd.md [hunk 4]
```diff
@@ -1,49 +1,163 @@
...
+you migrate your cluster e.g. from 2.2.x to 2.3.y, there is no way to get back
+to 2.2.x (other than restoring from backup data from the moment when 2.2.x was
+running, though if we were running 2.3.y for some time, we will lose all data
+that were written in the meantime).
+
+This is a very huge limitation - we really need a rollback procedure if we face
+some problems with the new release.
+
+To make things better, we were provided a custom rollback tool by CoreOS (etcd
```
Comment (xiang90) [inline docs/admin/etcd.md]: to be clear, this is a offline data rollback tool with very limited scope. it probably only works for v2 to v3 upgrades for k8s.

**Comment by ravilr** [comment]
@mml @wojtek-t can we also add some notes around how to deal with running pods in the cluster. I understand that the etcd migration is offline, during which kube-apiserver won't be able to talk to etcd. What happens to the already running pods in the cluster. Would the node controller evict nodes/pods due to lack of heartbeats as soon as the etcd is available but before the nodes starts heartbeating again. How to avoid affecting running pods in the cluster during this migration.

File: docs/admin/etcd.md [hunk 5]
```diff
@@ -1,49 +1,163 @@
...
+title: Upgrading and Rolling Back etcd
 ---
 
+## About etcd
+
 [etcd](https://coreos.com/etcd/docs/latest/) is a highly-available key value
 store which Kubernetes uses for persistent storage of all of its REST API
-objects.
+objects.  For the mechanics behind how kubernetes builds, distributes and
+deploys Kubernetes, see _some doc_.
```
Comment (ravilr) [inline docs/admin/etcd.md]: did you mean 'how kubernetes builds,.. deploys etcd'

File: docs/admin/etcd.md [hunk 6]
```diff
@@ -1,49 +1,163 @@
...
+The way etcd was designed introduces some significant limitations on how the
+upgrade can be done. These are the main limitations. 
+
+#### One minor release at a time
+
+It is possible to upgrade only by one minor release at a time (though, within
+patch releases you can switch versions at any time back and forth). That
+basically means, that we cannot upgrade directly e.g. from 2.1.x to 2.3.x.
+
+Fortunately, this limitation it is easy to workaround - it is enough to start a
```
Comment (justinsb) [inline docs/admin/etcd.md]: I think no comma after Fortunately, and s/it is/is/g

File: docs/admin/etcd.md [hunk 7]
```diff
@@ -1,49 +1,163 @@
...
+etcd in 2.2.z version, wait until it is healthy, stop it and then start then
+2.3.y version, wait until it is up and we are done.
+
+#### No rollback
+
+Etcd doesn’t support rollback procedure at all (by design). That means, that if
+you migrate your cluster e.g. from 2.2.x to 2.3.y, there is no way to get back
+to 2.2.x (other than restoring from backup data from the moment when 2.2.x was
+running, though if we were running 2.3.y for some time, we will lose all data
+that were written in the meantime).
```
Comment (justinsb) [inline docs/admin/etcd.md]: may be gramatically less controversial to say "all data written in the meantime"

File: docs/admin/etcd.md [hunk 8]
```diff
@@ -1,49 +1,163 @@
...
+
+This is a very huge limitation - we really need a rollback procedure if we face
```
Comment (justinsb) [inline docs/admin/etcd.md]: "very huge" is apt but perhaps not the bestest wording

File: docs/admin/etcd.md [hunk 9]
```diff
@@ -1,49 +1,163 @@
...
-## Kubernetes's usage of etcd
+### Migration script
+The migration script that will now be part of etcd docker image is a bash
+script, working as following:
 
-By default, Kubernetes objects are stored under the `/registry` key in etcd.
-This path can be prefixed by using the [kube-apiserver](/docs/admin/kube-apiserver) flag
-`--etcd-prefix="/foo"`.
+1. Detect which version of etcd we were previously running.
+   For that purpose, we have added a dedicated (metadata) file that will be
```
Comment (justinsb) [inline docs/admin/etcd.md]: This file is `version.txt` I believe - should we specify that?  (I don't have a strong opinion either way...)

File: docs/admin/etcd.md [hunk 10]
```diff
@@ -1,49 +1,163 @@
...
+   That means, for every minor etcd release between the detected one (excluding)
+   to the desired one (including):
+   1. start etcd in that version
+   1. wait until it is healthy (healthy means that you can write some data to
+      it)
+   1. stop this etcd
+   The important thing here is that those etcd will not list on the default etcd
+   port.  They are hardcoded to listen on ports which the apiserver is not
+   configured to connect to, which means that apiserver won’t be able to connect
+   to those etcd's and no new data will be written in the meantime.
```
Comment (justinsb) [inline docs/admin/etcd.md]: Although it _does_ remain part of the etcd cluster, because we don't change the peer port, I believe, so new data will be written.  Should we clarify the "no new data" statement?  (And is this actually safe?)

File: docs/admin/etcd.md [hunk 11]
```diff
@@ -1,49 +1,163 @@
...
+### Upgrade procedure
+Simply modify the command-line in etcd manifest to:
 
-```shell
-curl -fs -X PUT "http://${host}:${port}/v2/keys/_test"
-```
+1. Run the migration script
+   If the previously run version is already in the desired version, this will be
+   no-op.
+1. Start etcd (in the desired version)
```
Comment (justinsb) [inline docs/admin/etcd.md]: I think we should link to the salt manifest, and probably make this more explicit

i.e. 

Change "/usr/local/bin/etcd ..." to be
"if [ -e /usr/local/bin/migrate-if-needed.sh ]; then /usr/local/bin/migrate-if-needed.sh 1>>/var/log/etcd.log 2>&1; fi; /usr/local/bin/etcd..."

File: docs/admin/etcd.md [hunk 12]
```diff
@@ -1,49 +1,163 @@
...
-To test whether `etcd` is running correctly, you can try writing a value to a
-test key. On your master VM (or somewhere with firewalls configured such that
-you can talk to your cluster's etcd), try:
+### Upgrade procedure
+Simply modify the command-line in etcd manifest to:
 
-```shell
-curl -fs -X PUT "http://${host}:${port}/v2/keys/_test"
-```
+1. Run the migration script
```
Comment (justinsb) [inline docs/admin/etcd.md]: We also need:

* Specify the image as 3.0.17

* Specify these env vars for upgrade:

```
TARGET_STORAGE=etcd3
TARGET_VERSION=3.0.17
DATA_DIRECTORY=???
```

* Specify these env vars for downgrade (and clarify that the image should still be 3.0.17, as I made that mistake before engaging brain):

```
TARGET_STORAGE=etcd2
TARGET_VERSION=2.2.1
DATA_DIRECTORY=???
```

We also do specify this flag for etcd3 in salt; not sure if it matters: ` --quota-backend-bytes=4294967296`

**Comment by justinsb** [comment]
Great document - thank you!

Q: Is there any reason to stick with the 2.2.1 image?  My understanding is that the 3.0.17 image will still run etcd 2.2.1 if TARGET_STORAGE=="etcd2".  Sticking with one image would mean we wouldn't need to remember whether we were "etcd2 on a downgrade" or "etcd2 always", which would be great...

**Comment by justinsb** [comment]
We should also document that mixing versions of etcd-main and etcd-events does not work (unless there's some magic syntax for `etcd-servers-overrides`)

File: docs/admin/etcd.md (hunk 12 above)
Comment (liggitt) [inline docs/admin/etcd.md]: from https://github.com/kubernetes/kubernetes/issues/43366#issuecomment-287789004:

> [downgrading is] supported assuming that:
> 1. you are still using 'json' as the storage data format (even though the default is protobuf in 1.6)
> 2. you're not downgrading back to 2.2.1 image, but just using "3.0.17" image and setting ```ETCD_VERSION ``` to 2.2.1 

Documentation needs to make it crystal clear that you have to explicitly set `--storage-media-type=application/json` if you want to have the option to downgrade from etcd3 to etcd2.

It also needs to explain the "use the 3.0.17 image but run in 2.2.1 mode"... I'm not seeing that mentioned in this doc.

File: docs/admin/etcd.md [hunk 13]
```diff
@@ -1,49 +1,163 @@
...
+
+To make things better, we were provided a custom rollback tool by CoreOS (etcd
+folks), but:
+
+* This is a custom tool, which means it doesn’t have any guarantees and only
+  best-effort support from etcd team (though they fixed some bugs that we found
+  in the meantime).
+
+* The rollback can be done only from 3.0.x version (that is using v3 API) to
+  2.2.1 version (that is using v2 API).
```
Comment (liggitt) [inline docs/admin/etcd.md]: is this accurate? is that what @wojtek-t meant at https://github.com/kubernetes/kubernetes/issues/43366#issuecomment-287789004:

> you're not downgrading back to 2.2.1 image, but just using "3.0.17" image and setting ```ETCD_VERSION ``` to 2.2.1

Comment (liggitt) [inline docs/admin/etcd.md]: and only if all apiservers started against the etcd3 server ran with `--storage-media-type=application/json`, right?

File: docs/admin/etcd.md (hunk 10 above)
Comment (justinsb) [inline docs/admin/etcd.md]: I actually opened an issue about this before I knew about this doc: https://github.com/kubernetes/kubernetes/issues/43364

File: docs/admin/etcd.md [hunk 14]
```diff
@@ -1,49 +1,163 @@
...
+
+This is a very huge limitation - we really need a rollback procedure if we face
+some problems with the new release.
+
+To make things better, we were provided a custom rollback tool by CoreOS (etcd
+folks), but:
+
+* This is a custom tool, which means it doesn’t have any guarantees and only
+  best-effort support from etcd team (though they fixed some bugs that we found
+  in the meantime).
```
Comment (mml) [inline docs/admin/etcd.md]: Done.

File: docs/admin/etcd.md [hunk 15]
```diff
@@ -1,49 +1,163 @@
...
-to reduce downtime in case of corruption.
+### New etcd docker image
+We decided to completely change the way how etcd image looks like (and how it
+works). So far the docker image for etcd in version X contained only etcd and
+etcdctl binaries in version X.
 
-## Default configuration
+Going forward, the docker image for etcd in version X will contain multiple
+versions of etcd. For example 3.0.14 image will contain all: 2.2.1, 2.3.7 and
+3.0.14 binaries of etcd and etcdctl. This will allow running etcd in multiple
```
Comment (mml) [inline docs/admin/etcd.md]: Done.

File: docs/admin/etcd.md [hunk 16]
```diff
@@ -1,49 +1,163 @@
...
+watching (i.e.  they don’t send watch requests themselves). Both those
+frameworks if they can’t renew watch, they will start from “current version” by
+doing “list + watch from the resource version returned by list”. That means that
+if the apiserver will be down for the period of rollback, all of node components
+should basically restart their watches and start from “now” when apiserver is
+back. And it will be back with new resource version. That would mean that
+restarting node components is not needed.  But the assumptions here are pretty
+volatile…_
 
-## Configuration: high-level goals
```
Comment (mml) [inline docs/admin/etcd.md]: Done.  I made it a subsection of "About etcd" and renamed it.

File: docs/admin/etcd.md (hunk 1 above)
Comment (mml) [inline docs/admin/etcd.md]: Done :-).   But I will defer to whatever our style guide is about capitalizing words like `etcd` when they start a heading or sentence.

File: docs/admin/etcd.md (hunk 2 above)
Comment (mml) [inline docs/admin/etcd.md]: Would you prefer a different phrasing?  Regardless of why, the limitations are there.

File: docs/admin/etcd.md (hunk 5 above)
Comment (mml) [inline docs/admin/etcd.md]: Heh, fixed.

File: docs/admin/etcd.md (hunk 6 above)
Comment (mml) [inline docs/admin/etcd.md]: The comma belongs, but I'll defer to a "doc review" at some point.

See "sentence adverbs" at http://theeditorsblog.net/2016/02/21/a-tale-of-adverbs-and-the-comma/

I fixed the "it is".

File: docs/admin/etcd.md (hunk 7 above)
Comment (mml) [inline docs/admin/etcd.md]: Done.

File: docs/admin/etcd.md (hunk 8 above)
Comment (mml) [inline docs/admin/etcd.md]: changed to "significant"

File: docs/admin/etcd.md (hunk 9 above)
Comment (mml) [inline docs/admin/etcd.md]: Done.  I also tidied up the tense there to be more consistent, but I think it's still all over the map across the doc.

File: docs/admin/etcd.md (hunk 10 above)
Comment (mml) [inline docs/admin/etcd.md]: We are relying on the fact that the port we run on is obscure and "shouldn't" be exposed the way most masters are configured.  We should clarify that a bit.

If by safe, we mean we guarantee no one will try to write to it during this transition, then strictly speaking no.

File: docs/admin/etcd.md (hunk 12 above)
Comment (mml) [inline docs/admin/etcd.md]: Clarified this, I think.  PTAL.

I'd like to rearrange the doc so that the reader can get the recipe first and read the gory details about implementation later, if they wish.  Next revision, tho.

File: docs/admin/etcd.md (hunk 11 above)
Comment (mml) [inline docs/admin/etcd.md]: Where exactly are you suggesting I link to?

File: docs/admin/etcd.md (hunk 13 above)
Comment (mml) [inline docs/admin/etcd.md]: This is accurate, yes.  What @wojtek-t meant by that comment was that we don't need to change images.  The 3.0.17 image also contains the 2.2.1 binaries (and 2.3.7 and the rollback tool!).

File: docs/admin/etcd.md (hunk 1 above)
Comment (xiang90) [inline docs/admin/etcd.md]: Where is the rule about capitalizing words in the style guide? I cannot find it here: https://kubernetes.io/docs/contribute/style-guide/#documentation-formatting-standards.

**Comment by ethernetdan** [comment]
Has the feedback [here](https://github.com/kubernetes/kubernetes/issues/43366#issuecomment-287819980) been implemented? I'd like to close the issue.

File: docs/admin/etcd.md [hunk 17]
```diff
@@ -1,49 +1,203 @@
...
+
+Note that we need to migrate both: etcd versions that we are using (from 2.2.1
+to at least 3.0.x) as well as the version of API Kubernetes talks to etcd (3.0.x
+etcd binaries support both v2 and v3 API).
+
+This document describes how to do this migration.
+
+### etcd upgrade limitations
+
+The way etcd was designed introduces some significant limitations on how the
```
Comment (hongchaodeng) [inline docs/admin/etcd.md]: > The way etcd was designed introduces some significant limitations on how the upgrade can be done. These are the main limitations.

I disagree with the wording here. It sounds more like complaints than clarifying the limits to users.

Maybe rephrase like:
```
There are some limits in etcd upgrade process.
```

File: docs/admin/etcd.md [hunk 18]
```diff
@@ -1,49 +1,203 @@
...
+upgrade can be done. These are the main limitations.
+
+#### One minor release at a time
+
+It is possible to upgrade only by one minor release at a time (though, within
```
Comment (hongchaodeng) [inline docs/admin/etcd.md]: To emphasize and point out the upgrade path more explicitly, rephrase:
```
Upgrade only one minor release at a time. That means we cannot upgrade directly e.g. from 2.1.x to 2.3.x . Within patch releases, versions can be switched back and forth.
```

File: docs/admin/etcd.md [hunk 19]
```diff
@@ -1,49 +1,203 @@
...
+patch releases you can switch versions at any time back and forth). That
+basically means, that we cannot upgrade directly e.g. from 2.1.x to 2.3.x.
+
+Fortunately, this limitation is easy to workaround - it is enough to start a
```
Comment (hongchaodeng) [inline docs/admin/etcd.md]: Fortunately, ~~this limitation is easy to workaround~~

File: docs/admin/etcd.md [hunk 20]
```diff
@@ -1,49 +1,203 @@
...
+cluster for any intermediate minor release, wait until it is healthy and
+functional and stop it then. This will do the migration itself underneath.
+
+As an example: to upgrade from 2.1.x to 2.3.y version, it is enough to start
```
Comment (hongchaodeng) [inline docs/admin/etcd.md]: Combine this with above paragraph.

File: docs/admin/etcd.md [hunk 21]
```diff
@@ -1,49 +1,203 @@
...
+etcd in 2.2.z version, wait until it is healthy, stop it and then start then
+2.3.y version, wait until it is up and we are done.
+
+#### Rollback via special tool
+
+etcd versions through 3.0 don’t support general rollback. That is, in general
```
Comment (hongchaodeng) [inline docs/admin/etcd.md]: etcd versions ~~through 3.0~~ 3.0+

File: docs/admin/etcd.md [hunk 22]
```diff
@@ -1,49 +1,203 @@
...
+after migrating from M.N to M.N+1, there is no way to go back to M.N.
+
+This is a significant limitation - we really need a rollback procedure if we
```
Comment (hongchaodeng) [inline docs/admin/etcd.md]: Remove this paragraph. What can users get from reading this paragraph?

File: docs/admin/etcd.md [hunk 23]
```diff
@@ -1,49 +1,203 @@
...
+face some problems with the new release.
+
+To make things better, CoreOS has provided a
```
Comment (hongchaodeng) [inline docs/admin/etcd.md]: For the sake of open source spirit:

CoreOS -> etcd

File: docs/admin/etcd.md [hunk 24]
```diff
@@ -1,49 +1,203 @@
...
+restarting node components is not needed.  But the assumptions here are pretty
+volatile…_
 
-## Troubleshooting
+## Design
 
-To test whether `etcd` is running correctly, you can try writing a value to a
-test key. On your master VM (or somewhere with firewalls configured such that
-you can talk to your cluster's etcd), try:
+This section describes how we are going to do the migration given all the
```
Comment (hongchaodeng) [inline docs/admin/etcd.md]: ~~given all the limitations we have~~ given above conditions.

File: docs/admin/etcd.md [hunk 25]
```diff
@@ -1,49 +1,203 @@
...
+
+### Upgrade procedure
+Simply modify the command-line in etcd manifest to:
+
+1. Run the migration script
+   If the previously run version is already in the desired version, this will be
+   no-op.
+1. Start etcd (in the desired version)
+
+Starting in 1.6, this has been done in the manifests for new GCE clusters.  You
```
Comment (wojtek-t) [inline docs/admin/etcd.md]: Should this be put in the top?
@liggitt for input.

**Comment by wojtek-t** [comment]
> Has the feedback here been implemented? I'd like to close the issue kubernetes/kubernetes#43366

@ethernetdan - the modification to the doc, yes. Though there was also request to explicitly mention it in release note, and i think it's not yet done.

**Comment by jaredbhatti** [comment]
This appears to be a 1.6 release-bound item.  It's not in the tracking sheet (https://docs.google.com/spreadsheets/d/1nspIeRVNjAQHRslHQD1-6gPv99OcYZLMezrBe3Pfhhg/edit#gid=0) nor is it targeted against the 1.6 brach. Can you please fix these two issues?

**Comment by calebamiles** [comment]
made some small edits addressing your comments @hongchaodeng. PTAL

File: docs/admin/etcd.md [hunk 26]
```diff
@@ -1,49 +1,196 @@
...
+   version is communicated via environment variable), do the upgrade steps (if
+   needed).
+   That means, for every minor etcd release between the detected one (excluding)
+   to the desired one (including):
+   1. start etcd in that version
+   1. wait until it is healthy (healthy means that you can write some data to
+      it)
+   1. stop this etcd
+   The important thing here is that those etcd will not listen on the default
+   etcd port.  They are hardcoded to listen on ports which the apiserver is not
```
Comment (liggitt) [inline docs/admin/etcd.md]: are these ports secured identically to the normal startup?

File: docs/admin/etcd.md [hunk 27]
```diff
@@ -1,49 +1,196 @@
...
+Starting in 1.6, this has been done in the manifests for new GCE clusters.  You
+should also specify these environment variables.  In particular, **you must keep
+`STORAGE_MEDIA_TYPE` set to application/json if you wish to preserve the option
+to roll back**.
+
+```
+TARGET_STORAGE=etcd3
+ETCD_IMAGE=3.0.17
+TARGET_VERSION=3.0.17
+STORAGE_MEDIA_TYPE=application/json
```
Comment (liggitt) [inline docs/admin/etcd.md]: are these the GCE defaults or an example of what you would do to preserve the option to roll back?

File: docs/admin/etcd.md (hunk 26 above)
Comment (wojtek-t) [inline docs/admin/etcd.md]: These ports are just not exposed from the container to the outside world.

File: docs/admin/etcd.md (hunk 27 above)
Comment (wojtek-t) [inline docs/admin/etcd.md]: So this is what is enough to enable rollback in future. 

These are not OSS defaults.

File: docs/admin/etcd.md (hunk 11 above)
Comment (justinsb) [inline docs/admin/etcd.md]: I was suggesting to link to https://github.com/kubernetes/kubernetes/blob/master/cluster/saltbase/salt/etcd/etcd.manifest, as it may be easier to point to an example rather than saying it in words.

File: docs/admin/etcd.md [hunk 28]
```diff
@@ -1,49 +1,196 @@
...
+   to the desired one (including):
+   1. start etcd in that version
+   1. wait until it is healthy (healthy means that you can write some data to
+      it)
+   1. stop this etcd
+   The important thing here is that those etcd will not listen on the default
+   etcd port.  They are hardcoded to listen on ports which the apiserver is not
+   configured to connect to, which means that apiserver won’t be able to connect
+   to those etcd's.  Assuming no other client goes out of its way to try to
+   connect and write to this obscure port, no new data will be written during
```
Comment (justinsb) [inline docs/admin/etcd.md]: I remain unsure of what the procedure is for an HA cluster, trying to tease it out in https://github.com/kubernetes/kubernetes/issues/43364

Because we don't change the peer port AFAIK, other apiservers will write to their local etcds and these writes _will_ be replicated to the node being upgraded.  Even if not, in a 3 node cluster, 2 of the nodes will make progress while the first is being migrated.

It sounds like the procedure is to either put the apiserver into read-only mode or shut down the etcd cluster entirely (unsure as to how we're doing this in our e2e tests of this scenario - if anyone knows I'd like to follow the same procedure that we're testing).

Hopefully the upgrade works when the members were not in sync as well.

Comment (wojtek-t) [inline docs/admin/etcd.md]: Unfortunately the migration for not-synced masters is not supported by etcdctl migrate at all.
You can take a look here:
https://coreos.com/blog/migrating-applications-etcd-v3.html
Look into "Migration process" section.

So you need to ensure that in HA setup your instances are in sync.

**Comment by devin-donnelly** [comment]
@mml @wojtek-t Why are we changing the purpose of this document entirely instead of adding a Task or Tutorial for rolling back etcd? This doesn't seem like the appropriate course of action. There should be a document that explains etcd (the previous purpose for this page); just because etcd is changing doesn't necessarily mean that we repurpose the existing documentation to be about upgrade and rollback.

**Comment by mml** [comment]
@devin-donnelly I am fine making this a new doc and linking it in from the main etcd doc.  What do you think is the most appropriate way to do that?

I think as a separate section, it's going to be a little confusing, as in sheer volume these upgrade steps will dominate all the generic etcd information.

**Comment by calebamiles** [comment]
@devin-donnelly, @mml  I think when combined with https://github.com/kubernetes/kubernetes.github.io/pull/2898 that rewriting the current doc is somewhat appropriate. The level of detail of the current documentation at `admin/etcd.md` isn't really useful for cluster operators. Given the release of etcdv3 I can imagine that the migration and rollback process is far and away the most pressing operational concern that a cluster operator would have with respect to etcd. Additionally the `Concepts` section could probably be improved to link to the upstream etcd documents or be expanded with more generally information about etcd but I'm entirely willing to defer to your better judgement about documentation

**Comment by mml** [comment]
@jaredbhatti rebased and retargeted.  I added this PR to the spreadsheet.  There are now two PRs in the row.

**Comment by mml** [comment]
> Q: Is there any reason to stick with the 2.2.1 image? My understanding is that the 3.0.17 image will still run etcd 2.2.1 if TARGET_STORAGE=="etcd2". Sticking with one image would mean we wouldn't need to remember whether we were "etcd2 on a downgrade" or "etcd2 always", which would be great...

The 2.2.1 image only has one advantage: you can't misconfigure it with nonsensical combinations of environment variables (like `TARGET_STORAGE=etcd, TARGET_VERSION=2.2.1`).  But if configured correctly, the 3.0.17 image does everything the 2.2.1 one does and more.

**Comment by mml** [comment]
> We should also document that mixing versions of etcd-main and etcd-events does not work (unless there's some magic syntax for etcd-servers-overrides)

@justinsb Where would you suggest I call that out?

**Comment by mml** [comment]
@calebamiles #2898 is not targeted for 1.6, so in the interim the current doc will disappear.  Is that cool?

File: docs/admin/etcd.md [hunk 29]
```diff
@@ -1,49 +1,196 @@
...
+### etcd upgrade requirements
+
+There are requirements on how an etcd cluster upgrade can be performed; the primary considerations are
+- [upgrade between one minor release at a time](#one-minor-release-at-a-time)
+- [rollback supported through additional tooling](#rollback-via-additional-tooling)
+
+#### One minor release at a time
+
+Upgrade only one minor release at a time, e.g. we cannot upgrade directly from 2.1.x to 2.3.x.
+Within patch releases it is possible to upgrade and downgrade between arbitrary versions.It is fairly
```
Comment (wojtek-t) [inline docs/admin/etcd.md]: I'm not native speaker, but for me something is missing before "It is fairly easy...". I'm missing some context for this, sth like:
"Fortunately, just starting a cluster for any intermediate minor release, waiting until it is healthy and functional and then shutting it down is itself performing the migration."

WDYT?

File: docs/admin/etcd.md [hunk 30]
```diff
@@ -1,49 +1,196 @@
...
+but:
+
+* This custom tool is not part of the etcd repo and does not receive the same
+  testing as the rest of etcd.  We are testing it in a couple of e2e tests.
+  There is only community support here.
+
+* The rollback can be done only from 3.0.x version (that is using v3 API) to
+  2.2.1 version (that is using v2 API).
+
+* The tool only works if the data is still stored in `application/json` format.
```
Comment (wojtek-t) [inline docs/admin/etcd.md]: Make it bold?

**Comment by wojtek-t** [comment]
LGTM - but I guess we also need LGTM signal from tech writer.
@devin-donnelly ?

**Comment by wojtek-t** [comment]
Also - please squash commits.

File: docs/admin/etcd.md [hunk 31]
```diff
@@ -1,49 +1,196 @@
 ---
```
Comment (xiang90) [inline docs/admin/etcd.md]: @mml @wojtek-t can we change the name of the md file to `etcd_upgrade.md` to match the title?

I am going to revive the original generic etcd doc soon, which should be named as etcd.md.

Comment (wojtek-t) [inline docs/admin/etcd.md]: I'm not good at naming - I'm fine with that, but will leave that decision to others.

Comment (calebamiles) [inline docs/admin/etcd.md]: I'll add the original etcd.md to this PR @xiang90 and I can make some small changes to update that doc if you'd like

File: docs/admin/etcd.md [hunk 32]
```diff
@@ -1,49 +1,196 @@
...
+title: Upgrading and Rolling Back etcd
 ---
 
+## About etcd
+
 [etcd](https://coreos.com/etcd/docs/latest/) is a highly-available key value
 store which Kubernetes uses for persistent storage of all of its REST API
-objects.
+objects.  For the mechanics behind how kubernetes builds, distributes and
+deploys etcd, see _some doc_.
```
Comment (chenopis) [inline docs/admin/etcd.md]: Is _some doc_ just a placeholder for a link? If so, this needs to put in before we can Docs LGTM.

Comment (mml) [inline docs/admin/etcd.md]: It is a placeholder.  I'll just take the sentence out until _some doc_ is ready.  I do not want to hold up this PR any further.

**Comment by steveperry-53** [comment]
@mml, @wojtek-t, See https://github.com/kubernetes/kubernetes.github.io/pull/2767/commits/cb071680a2090626f89d931098ea4810b41c1c86 for suggested edits.

File: docs/admin/etcd_upgrade.md [hunk 33]
```diff
@@ -0,0 +1,209 @@
...
+* Rollback doesn’t preserve resource versions of objects stored in etcd.
+
+The last bullet means that any component or user that has some logic
+depending on resource versions may require restart after etcd rollback. This
+includes that all clients using the watch API, which depends on
+resource versions. Since both the kubelet and kube-proxy use the watch API, a
+rollback might require restarting all Kubernetes components on all nodes.
+
+_To verify: It seems that both Kubelet and KubeProxy are using “resource
+version” only for watching (i.e. are not using resource versions for anything
```
Comment (steveperry-53) [inline docs/admin/etcd_upgrade.md]: Is this text under "To verify" meant to be in the document? It almost looks like a review comment.

Comment (mml) [inline docs/admin/etcd_upgrade.md]: Good call.  I will rephrase that.

**Comment by mml** [comment]
@steveperry-53 edits mostly look good, but some of the seemingly excessive bolding is because there are some mistakes users can make here that could cut off their option for rollback or that would lead them into waters where we have not tested.  These are charted waters, but the chart is not the territory.

So I guess the question is one of doc style.  How do we warn users about something risky and important?  This is like those sidebars usually indicated by something like `(!)` in "for dummies" books.

Thanks!

**Comment by steveperry-53** [comment]
@mml, If you feel it's important, go ahead and return some important statements to bold. Usually what we do, instead of bold, is to separate important material as a Note or Warning. We bold the word **Note** or **Warning** but not the text. [Here](https://kubernetes.io/docs/concepts/cluster-administration/sysctl-cluster/) is an example.

**Comment by mml** [comment]
@steveperry-53 Thanks, I'll try to restructure things to follow that style.  I am re-labeling do-not-merge until I finish those changes and squash.

File: docs/admin/etcd.md (hunk 30 above)
Comment (mml) [inline docs/admin/etcd.md]: This is now called out as a **warning** per @steveperry-53's style guidance.

**Comment by mml** [comment]
We are ready to go.

**Comment by steveperry-53** [comment]
@mml, Would you like me to merge this now?

**Comment by mml** [comment]
@steveperry-53 yes, pleas.

File: docs/admin/etcd_upgrade.md [hunk 34]
```diff
@@ -0,0 +1,214 @@
...
+This section describes how we are going to do the migration, given the
+[etcd upgrade requirements](#etcd-upgrade-requirements).
+
+Note that because the code changes in Kubernetes code needed
+to support the etcd v3 API are local and straightforward, we do not
+focus on them at all. We focus only on the upgrade/rollback here.
+
+### New etcd Docker image
+
+We decided to completely change the the content of the etcd image and the way it works.
```
Comment (jagosan) [inline docs/admin/etcd_upgrade.md]: typo: "the the"
